torchtext/experimental/datasets/text_classification.py

```diff
@@ -1,93 +1,455 @@
 
-def _imdb_iterator(key, extracted_files, tokenizer, ngrams, yield_cls=False):
-    for fname in extracted_files:
-        if 'urls' in fname:
-            continue
-        elif key in fname and ('pos' in fname or 'neg' in fname):
-            with io.open(fname, encoding="utf8") as f:
-                label = 1 if 'pos' in fname else 0
-                if yield_cls:
-                    yield label, ngrams_iterator(tokenizer(f.read()), ngrams)
-                else:
-                    yield ngrams_iterator(tokenizer(f.read()), ngrams)
+    def __len__(self):
+        return len(self.data)
 
+    def get_labels(self):
+        return set([self.transforms[0](item[0]) for item in self.data])
 
-def _generate_data_iterators(dataset_name, root, ngrams, tokenizer, data_select):
+    def get_vocab(self):
+        return self.vocab
+
+
+def _setup_datasets(dataset_name, root='.data', ngrams=1, vocab=None,
+                    tokenizer=None, data_select=('train', 'test')):
     if not tokenizer:
-        tokenizer = get_tokenizer("basic_english")
+        tok_transform = TokenizerTransform()
+    else:
+        tok_transform = TokenizerTransform(tokenizer)
 
     if not set(data_select).issubset(set(('train', 'test'))):
         raise TypeError('Given data selection {} is not supported!'.format(data_select))
 
-    dataset_tar = download_from_url(URLS[dataset_name], root=root)
-    extracted_files = extract_archive(dataset_tar)
+    ngram_transform = NGrams(ngrams)
+    processing_transform = TextSequential(tok_transform, ngram_transform)
 
-    iters_group = {}
-    if 'train' in data_select:
-        iters_group['vocab'] = _imdb_iterator('train', extracted_files,
-                                              tokenizer, ngrams)
-    for item in data_select:
-        iters_group[item] = _imdb_iterator(item, extracted_files,
-                                           tokenizer, ngrams, yield_cls=True)
-    return iters_group
+    train, test = DATASETS[dataset_name](root=root)
+    raw_data = {'train': train,
+                'test': test}
 
+    if not vocab:
+        if 'train' not in data_select:
+            raise TypeError("Must pass a vocab if train is not selected.")
+        vocab = build_vocab(train, processing_transform)
+    label_transform = ToTensor(dtype=torch.long)
+    text_transform = TextSequential(processing_transform,
+                                    VocabTransform(vocab),
+                                    ToTensor(dtype=torch.long))
+    return tuple(TextClassificationDataset(raw_data[item], vocab,
+                                           (label_transform, text_transform))
+                 for item in data_select)
 
-def _setup_datasets(dataset_name, root='.data', ngrams=1, vocab=None,
-                    removed_tokens=[], tokenizer=None,
-                    data_select=('train', 'test')):
 
-    if isinstance(data_select, str):
-        data_select = [data_select]
+def AG_NEWS(*args, **kwargs):
```

Comment:
If someone does "help(AG_NEWS())"  will it render the arguments correctly? Because of the forwarding this might not work.